### ISSUE TITLE
Prompt-completion does not use correct labels when passing conversational format to SFT

### ISSUE THREAD

**Issue by MarkusSpanring** [huggingface/trl#4147]
### Reproduction

Hi!

With the switch to the new dataset format types for SFT I was wondering how to properly provide the data for training, in my case prompt-completion
I have used `DataCollatorForCompletionOnlyLM` in the past and now want to update to the newest trl version to get all the nice new features.

In the past, I had my raw data in the message format:
```
# example 1
example = {
    "messages": [
        {"role": "system", "content": "SYSTEM"},
        {"role": "user", "content": "QUESTION"},
        {"role": "assistant", "content": "ANSWER"},
    ]
}
```
And I applied the chat template on it, which gave me something like this (generic chat tempalte):
```
<system>SYSTEM<end><user>QUESTION<end><assistant>ANSWER<end>
```
With the `DataCollatorForCompletionOnlyLM` I had to specify the instruction and response templates as `<user>` and `<assistant>` respectively and it then correctly gave me the labels and masks.

Now with the new format (as described [here](https://huggingface.co/docs/trl/v0.23.0/en/dataset_formats#conversational)) I need to bring my data in the `prompt-completion` format.
```
# example 2
example = {
    "prompt": [
        {"role": "system", "content": "SYSTEM"},
        {"role": "user", "content": "QUESTION"}
    ],
    "completion": [
        {"role": "assistant", "content": "ANSWER"}
    ]
}
```

Now to my question: How do I pass these examples to SFT?

Naively, I would have assumed that I can just pass the examples in the role-content format to SFT and the chat template is applied accordingly.
However, when inspecting the tokenized inputs and labels I see the following
```
Input: <system>SYSTEM<end><user>QUESTION<end><assistant>ANSWER<end>
Label: --------------------------------------<assistant>ANSWER<end>
```
I.e. the generation prompt is added in the labels and not in the prompt.

When I process example 2 with `apply_chat_template` before (as recommended [here](https://huggingface.co/docs/trl/v0.23.0/en/dataset_formats#working-with-conversational-datasets-in-trl)), the example looks as follows:
```
# processed example 2
example = {
    "prompt": "<system>SYSTEM<end><user>QUESTION<end><assistant>",
    "completion": "ANSWER<end>"
}
```

And the corresponding input and label is as I would have expected it 
```
Input: <system>SYSTEM<end><user>QUESTION<end><assistant>ANSWER<end>
Label: -------------------------------------------------ANSWER<end>
```

Comparing how the chat template is applied in `apply_chat_template` and `_prepare_dataset` ([here](https://github.com/huggingface/trl/blob/main/trl/trainer/sft_trainer.py#L968)) the generation prompt is added in the former since the last role in `prompt` is naturally `user`.

I am leaning towards saying that this is a bug in `_prepare_dataset` but could also be that I have misunderstood how to use the new formats.

Any help is very much appreciated!

Here are the minimal examples:
```
# Passing Conversational format directly
from trl import SFTTrainer
from datasets import load_dataset
from datasets import Dataset

example = Dataset.from_dict({
    "prompt": [[
        {"role": "system", "content": "SYSTEM"},
        {"role": "user", "content": "QUESTION"}
    ]],
    "completion": [[
        {"role": "assistant", "content": "ANSWER"}
    ]]
})

trainer = SFTTrainer(
    model="Qwen/Qwen3-0.6B",
    train_dataset=example,
)

tokenized_data = trainer.data_collator.torch_call(trainer.train_dataset)
print(repr(trainer.processing_class.decode(tokenized_data["input_ids"][0].tolist())))
print(repr(trainer.processing_class.decode([i for i in tokenized_data["labels"][0].tolist() if i != -100])))
```

```
# Applying chat template before with apply_chat_template and passing standard format
from datasets import Dataset
from trl import apply_chat_template
from transformers import AutoTokenizer
from trl import SFTTrainer

example = Dataset.from_dict({
    "prompt": [[
        {"role": "system", "content": "SYSTEM"},
        {"role": "user", "content": "QUESTION"}
    ]],
    "completion": [[
        {"role": "assistant", "content": "ANSWER"}
    ]]
})


tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-0.6B")
example = example.map(apply_chat_template, fn_kwargs={"tokenizer": tokenizer})

trainer = SFTTrainer(
    model="Qwen/Qwen3-0.6B",
    train_dataset=example,
)

tokenized_data = trainer.data_collator.torch_call(trainer.train_dataset)
print(repr(trainer.processing_class.decode(tokenized_data["input_ids"][0].tolist())))
print(repr(trainer.processing_class.decode([i for i in tokenized_data["labels"][0].tolist() if i != -100])))
```

### System Info

- Platform: Linux-6.8.0-1031-aws-x86_64-with-glibc2.35
- Python version: 3.12.10
- TRL version: 0.23.0
- PyTorch version: 2.8.0
- accelerator(s): NVIDIA A10G
- Transformers version: 4.56.2
- Accelerate version: 1.10.1
- Accelerate config: not found
- Datasets version: 3.6.0
- HF Hub version: 0.34.4
- bitsandbytes version: 0.46.1
- DeepSpeed version: not installed
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.95.1
- PEFT version: 0.16.0
- vLLM version: not installed

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by behroozazarkhalili** [comment]
Fix for GitHub Issue #4147: Prompt-Completion Labeling Bug

## Problem
When using prompt-completion dataset format with conversational data in SFTTrainer, the labels were incorrectly generated. Specifically:

1. When `assistant_only_loss=True` and chat template doesn't support `{% generation %}` keyword
2. The `assistant_masks` returned by `apply_chat_template` are all zeros (invalid)
3. The data collator would use these invalid masks, masking ALL tokens including the completion
4. Result: No training happens because all labels are set to -100

## Root Cause
In `sft_trainer.py` lines 937-983, the code:
1. Tokenizes incomplete conversation (prompt only)
2. Tokenizes complete conversation (prompt + completion)
3. Creates `completion_mask` based on length difference
4. Also gets `assistant_masks` from `apply_chat_template`

The bug: When `assistant_masks` exists but is invalid (all zeros), it still gets included in the output. The data collator (lines 216-220) applies both masks sequentially, with `assistant_masks` overwriting the correct `completion_mask`, resulting in all labels being -100.

## Solution
Modified `sft_trainer.py` lines 979-989:

```python
# Handle assistant_masks from apply_chat_template
# If assistant_masks exists but is invalid (all zeros), remove it and use completion_mask instead
# This happens when chat template doesn't support generation tokens (no {% generation %})
if "assistant_masks" in output and 1 not in output["assistant_masks"]:
    del output["assistant_masks"]

# Create completion_mask if assistant_masks is not available
# assistant_masks from apply_chat_template is more reliable when available
if "assistant_masks" not in output:
    completion_mask = [0] * len(prompt_ids) + [1] * (len(prompt_completion_ids) - len(prompt_ids))
    output["completion_mask"] = completion_mask
```

## Behavior After Fix

### Case 1: Chat template supports generation tokens
- `assistant_masks` has valid values (contains 1s)
- Use `assistant_masks` for labeling ✅

### Case 2: Chat template doesn't support generation tokens
- `assistant_masks` returned but all zeros (invalid)
- Delete invalid `assistant_masks` from output
- Use `completion_mask` instead ✅

### Case 3: No assistant_only_loss flag
- No `assistant_masks` requested
- Use `completion_mask` ✅

## Test Results

### Comprehensive Testing with All Issue Examples

#### Test 1: Example 1 - Messages Format (Language Modeling)
**Input:**
```python
{
    "messages": [
        {"role": "system", "content": "SYSTEM"},
        {"role": "user", "content": "QUESTION"},
        {"role": "assistant", "content": "ANSWER"}
    ]
}
```

**Result:** ✅ Works correctly (trains on full sequence as expected)

#### Test 2: Example 2 - Prompt-Completion Format (THE BUG CASE)
**Input:**
```python
{
    "prompt": [
        {"role": "system", "content": "SYSTEM"},
        {"role": "user", "content": "QUESTION"}
    ],
    "completion": [
        {"role": "assistant", "content": "ANSWER"}
    ]
}
```


**Before Fix:** Labels were all -100 (no training) or included generation prompt ❌
**After Fix:** Labels = `'<|im_start|>assistant\nANSWER<|im_end|>\n'` ✅
- Prompt tokens correctly masked (not trained on) ✅
- Completion tokens correctly trained on ✅

#### Test 3: Minimal Reproduction (From Issue)
**Result:** ✅ Labels correctly contain only the completion

#### Test 4: Comparison of completion_only_loss vs assistant_only_loss
**Both modes:** Produce identical correct labels for prompt-completion format ✅

### Summary of Test with Qwen/Qwen2.5-0.5B-Instruct

**Prompt-Completion Format:**
- `completion_mask` correctly created: `[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1]`
- Training tokens: `<|im_start|>assistant\nANSWER<|im_end|>`
- Invalid `assistant_masks` (all zeros) correctly removed
- Fallback to `completion_mask` works perfectly

## Status
✅ **Fixed and comprehensively tested - All examples from issue #4147 pass successfully**

**Comment by behroozazarkhalili** [comment]
@MarkusSpanring, could you please check if this fix resolves the issue you encountered?
@qgallouedec, if you confirm the above fix, I can open a pull request PR to update `SFTTrainer`.

**Comment by qgallouedec** [comment]
I'm not sure to understand @behroozazarkhalili, actually I think it's because `add_generation_prompt=True` is missing here

https://github.com/huggingface/trl/blob/main/trl/trainer/sft_trainer.py#L938-L939

It's a pretty important issue if I'm correct, I'd love a PR to fix it

**Comment by behroozazarkhalili** [comment]
@qgallouedec You're absolutely right that `add_generation_prompt=True` is the key insight! However, your fix alone doesn't solve the issue. Let me explain why.

## The Root Cause

The bug happens in the **data collator**, not in the tokenization. The data collator applies masks in this order:

```python
# DataCollatorForLanguageModeling (lines 211-220)
if self.completion_only_loss and "completion_mask" in examples[0]:
    output["labels"][completion_mask == 0] = -100  # Step 1

if "assistant_masks" in examples[0]:
    output["labels"][assistant_masks == 0] = -100  # Step 2 - OVERWRITES Step 1!
```

When `assistant_only_loss=True` and the chat template lacks `{% generation %}` support, `assistant_masks` returns **all zeros**. These invalid masks **overwrite** the correct `completion_mask`, masking everything.

## Test Evidence

**Tokenizer:** `Qwen/Qwen2.5-0.5B-Instruct`

```python
prompt = [
    {"role": "system", "content": "SYSTEM"},
    {"role": "user", "content": "QUESTION"}
]
completion = [
    {"role": "assistant", "content": "ANSWER"}
]
```

### Scenario 1: Only your fix (add_generation_prompt=True)

```python
# Line 940: Add generation prompt
prompt_ids = tokenizer.apply_chat_template(
    prompt,
    tokenize=True,
    add_generation_prompt=True  # ← Your fix
)

# Result: Better masking precision
# Prompt: 15 tokens (includes <|im_start|>assistant)
# completion_mask: [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, 1,1,1,1]  ✓ Correct! 4 tokens

# But assistant_masks is still requested when assistant_only_loss=True
full = tokenizer.apply_chat_template(
    prompt + completion,
    return_assistant_tokens_mask=True
)
# assistant_masks: [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]  ✗ All zeros!
```

**What happens in the data collator:**
```
Step 1: completion_mask sets labels for 4 tokens  ✓
Step 2: assistant_masks (all zeros) overwrites EVERYTHING with -100  ✗

Result: 0 training tokens  ❌ FAILS
```

### Scenario 2: Both fixes combined

```python
# Your fix: Better precision
prompt_ids = tokenizer.apply_chat_template(
    prompt,
    tokenize=True,
    add_generation_prompt=True
)

# My fix: Remove invalid masks
if "assistant_masks" in output and 1 not in output["assistant_masks"]:
    del output["assistant_masks"]  # ← Prevents overwrite

if "assistant_masks" not in output:
    completion_mask = [0] * len(prompt_ids) + [1] * (len(full_ids) - len(prompt_ids))
```

**What happens in the data collator:**
```
Step 1: completion_mask sets labels for 4 tokens  ✓
Step 2: assistant_masks not present, skip  ✓

Result: 4 training tokens = 'ANSWER<|im_end|>\n'  ✅ WORKS
```

## Why Your Fix Alone Isn't Enough

Even with `add_generation_prompt=True`:
1. ✅ Correct `completion_mask`
3. ❌ But `assistant_masks` are still all zeros
4. ❌ Data collator still overwrites everything with -100

## The Complete Solution

**Your contribution:** `add_generation_prompt=True` → Achieves precise masking
**My contribution:** Remove invalid `assistant_masks` → Prevents overwrite

Both are needed:
- Without your fix: Fails (7 tokens instead of 4)
- Without my fix: Fails completely (0 tokens, no training)
- With both: Works perfectly (4 tokens)

## Proposed Implementation

```python
# Line 940: Your fix
prompt_ids = processing_class.apply_chat_template(
    example["prompt"],
    tokenize=True,
    add_generation_prompt=True,  # ← @qgallouedec
    tools=example.get("tools"),
    **example.get("chat_template_kwargs", {}),
)

# Lines 980-989: My fix
if "assistant_masks" in output and 1 not in output["assistant_masks"]:
    del output["assistant_masks"]  # ← @behroozazarkhalili

if "assistant_masks" not in output:
    completion_mask = [0] * len(prompt_ids) + [1] * (len(prompt_completion_ids) - len(prompt_ids))
    output["completion_mask"] = completion_mask
```

Both fixes are essential for the complete solution. What do you think?
I'll add a pull request when you confirm.

**Comment by qgallouedec** [comment]
I think the assistant mask "issue" is something different. Indeed you shouldn't use `assistant_only_loss` when your model's chat template doesn't support it. It's clearly stated in the doc but maybe we can raise a warning when the returned assistant mask is all zero, to make it even clearer. However, I wouldn't force assistant mask to be ignored when all zero, because it would mean silently ignore the user setup.

**Comment by behroozazarkhalili** [comment]
@qgallouedec You're absolutely right! I've updated the implementation to follow your strict approach:

## What I Implement

**Only your fix + warning (no silent fallback):**

```python
# Line 940: add_generation_prompt=True for precision
prompt_ids = processing_class.apply_chat_template(
    example["prompt"],
    tokenize=True,
    add_generation_prompt=True,  # ← More precise masking
    tools=example.get("tools"),
    **example.get("chat_template_kwargs", {}),
)

# Lines 980-988: Warning when assistant_masks are all zeros
if "assistant_masks" in output and 1 not in output["assistant_masks"]:
    logger.warning(
        "The chat template does not support generation tokens (missing '{% generation %}' keyword). "
        "When using prompt-completion format with assistant_only_loss=True, this will result in "
        "all tokens being masked. Consider using a chat template that supports generation tokens, "
        "or set assistant_only_loss=False and use completion_only_loss=True instead."
    )
```

**No `completion_mask` fallback** - users must fix their setup.

This forces users to either:
1. Use a chat template supporting `{% generation %}`
2. Switch to `completion_only_loss=True` instead of `assistant_only_loss=True`

Is this the approach you wanted?

**Comment by MarkusSpanring** [comment]
@behroozazarkhalili @qgallouedec do you still need me to verify the behaviour? From the comments above I take that root cause for the issue was found and that the implemented fix will lead to properly masked labels for both `completion_only_loss` and `assistant_only_loss`. I am aware that the second will only work with chat templates that support `{% generation %}`.

**Comment by behroozazarkhalili** [comment]
@MarkusSpanring, we will implement @qgallouedec's solution and issue a warning for the other case.

**Comment by qgallouedec** [comment]
@MarkusSpanring can you try adding `add_generation_prompt=True` here


https://github.com/huggingface/trl/blob/1eff7da9e0fe75106111ddb15934f6395d148bf2/trl/trainer/sft_trainer.py#L937-L942

and see if it solves your issue?

**Comment by MarkusSpanring** [comment]
@qgallouedec as far as I can say this solves the issue, yes.
